### PR TITLE
Implement blocking inlining for kwcall with rule

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

File: lib/EnzymeCore/test/runtests.jl [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+@test has_frule_from_sig(Base.signature_type(g, Tuple{Float64}))
+
+f(;kwargs) = 1.0
+
+function forward(::Const{typeof(f)}, ::Type{<:Const}; kwargs...)
+    return Const(f(; kwargs...))
+end
+
+@test has_frule_from_sig(Base.signature_type(f, Tuple{}))
```
Comment (wsmoses) [inline lib/EnzymeCore/test/runtests.jl:19]: Is it useful to test where we have a kwarg function but a rule that doesn't take the kwarg?

**Review by wsmoses — APPROVED** [review]

File: lib/EnzymeCore/test/runtests.jl (hunk 1 above)
Comment (vchuravy) [reply lib/EnzymeCore/test/runtests.jl:19]: Yeah we will error for Enzyme proper (at least that's the goal)

**Comment by vchuravy** [comment]
@wsmoses 

```
using Enzyme
using Enzyme.EnzymeRules

import .EnzymeRules: forward

function f_kw(x; kwargs...)
    x^2
end

function forward(::Const{typeof(f_kw)}, ::Type{<:DuplicatedNoNeed}, x::Duplicated; kwargs...)
    # @show kwargs
    return 10+2*x.val*x.dval
end

@show autodiff(Forward, f_kw, Duplicated(2.0, 1.0))[1] ≈ 14.0

# TODO: autodiff wrapper with kwargs support

g(x, y) = f_kw(x; val=y)
@show autodiff(Forward, g, Duplicated(2.0, 1.0), Const(42.0))[1] ≈ 14.0
```

Can you look at the calling convention mismatch?

**Comment by vchuravy** [comment]
```
JULIA_DEBUG=Enzyme julia --project -L kwtest.jl
[ Info: Precompiling Enzyme [7da242da-08ed-463a-9acd-ee780be4f1d9]
┌ Debug: Found frule for
│   mi.specTypes = Tuple{typeof(f_kw), Float64}
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:50
┌ Debug: Trying to apply custom forward rule
│   TT = Tuple{Const{typeof(f_kw)}, Type{DuplicatedNoNeed{Float64}}, Duplicated{Float64}}
│   isKWCall = false
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:50
┌ Debug: Applying custom forward rule
│   TT = Tuple{Const{typeof(f_kw)}, Type{DuplicatedNoNeed{Float64}}, Duplicated{Float64}}
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:50
(autodiff(Forward, f_kw, Duplicated(2.0, 1.0)))[1] ≈ 14.0 = true
┌ Debug: Blocking inlining due to frule
│   mi.specTypes = Tuple{typeof(Core.kwcall), NamedTuple{(:val,), Tuple{Float64}}, typeof(f_kw), Float64}
└ @ Enzyme.Compiler.Interpreter ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:50
┌ Debug: Found frule for
│   mi.specTypes = Tuple{typeof(Core.kwcall), NamedTuple{(:val,), Tuple{Float64}}, typeof(f_kw), Float64}
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:50
┌ Debug: Trying to apply custom forward rule
│   TT = Tuple{NamedTuple{(:val,), Tuple{Float64}}, typeof(forward), Const{typeof(f_kw)}, Type{DuplicatedNoNeed{Float64}}, Duplicated{Float64}}
│   isKWCall = true
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:50
┌ Debug: Applying custom forward rule (kwcall)
│   TT = Tuple{NamedTuple{(:val,), Tuple{Float64}}, typeof(forward), Const{typeof(f_kw)}, Type{DuplicatedNoNeed{Float64}}, Duplicated{Float64}}
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:50
┌ Error: ("Calling convention mismatch", [1 x double] addrspace(11)*,   %10 = addrspacecast [1 x [1 x double]] addrspace(10)* %9 to [1 x [1 x double]] addrspace(11)*, !dbg !9, 1, define double @julia_kwcall_1979([1 x double] addrspace(11)* nocapture nofree noundef nonnull readnone align 8 dereferenceable(8) %0, [2 x double] addrspace(11)* nocapture nofree noundef nonnull readonly align 8 dereferenceable(16) %1) #6 !dbg !35 {
│ top:
│   %2 = call {}*** @julia.get_pgcstack()
│   %3 = getelementptr inbounds [2 x double], [2 x double] addrspace(11)* %1, i64 0, i64 0, !dbg !36
│   %4 = getelementptr inbounds [2 x double], [2 x double] addrspace(11)* %1, i64 0, i64 1, !dbg !36
│   %5 = load double, double addrspace(11)* %3, align 8, !dbg !42, !tbaa !50, !invariant.load !8, !alias.scope !54, !noalias !57
│   %6 = fmul double %5, 2.000000e+00, !dbg !42
│   %7 = load double, double addrspace(11)* %4, align 8, !dbg !62, !tbaa !50, !invariant.load !8, !alias.scope !54, !noalias !57
│   %8 = fmul double %6, %7, !dbg !62
│   %9 = fadd double %8, 1.000000e+01, !dbg !63
│   ret double %9, !dbg !41
│ }
│ )
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:50
```

**Comment by wsmoses** [comment]
I think the code around here will need help: https://github.com/EnzymeAD/Enzyme.jl/blob/f3d2868c6f17532cefd015a12020fe082e40e11d/src/compiler.jl#L3119

**Comment by wsmoses** [comment]
@vchuravy I fixed your test to work properly now. I will note something: currently we assume that both the original funnction, and custom rule have the same kwarg behavior (eg both have or both don't have). Handling the other combos is also likely to be necessary.

**Comment by vchuravy** [comment]
Thanks! 

With the other combinations you mean:
- Primal has keyword args, but rule is missing them
- Primal doesn't have keyword arg, but rule has them 

The first should error, the second one is fine.

We will need to test the case where the rule has different kwargs

**Comment by wsmoses** [comment]
Can we have a test that tests the "active kwarg not handled". Say by returning x*val?

**Comment by vchuravy** [comment]
@vtjnash do you have a hint for how we could implement kwfunc detection on 1.7 & 1.8?

File: src/compiler/interpreter.jl [hunk 2]
```diff
@@ -167,18 +201,21 @@ elseif isdefined(Core.Compiler, :is_stmt_inline)
 function Core.Compiler.inlining_policy(interp::EnzymeInterpreter,
     @nospecialize(src), stmt_flag::UInt8, mi::MethodInstance, argtypes::Vector{Any})
 
-    if is_primitive_func(mi.specTypes)
+    @safe_debug "Inspecting... " mi.specTypes
```
Comment (vchuravy) [inline src/compiler/interpreter.jl:204]:
```suggestion
```

File: src/compiler.jl [hunk 3]
```diff
@@ -3699,15 +3814,17 @@ function arraycopy_fwd(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef
...
             end
         end
-     
+
         unsafe_store!(shadowR, shadowres.ref)
     end
-	
+
+    unsafe_store!(shadowR, shadowres.ref)
+
```
Comment (vchuravy) [inline src/compiler.jl:3822]:
```suggestion
```

File: src/compiler.jl [hunk 4]
```diff
@@ -3840,39 +3957,40 @@ function arraycopy_common(fwd, B, orig, origArg, gutils, shadowdst)
...
 	return nothing
 end
 
 function arraycopy_rev(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef, gutils::API.EnzymeGradientUtilsRef, tape::LLVM.API.LLVMValueRef)::Cvoid
     orig = LLVM.Instruction(OrigCI)
     origops = LLVM.operands(orig)
-    
-    if API.EnzymeGradientUtilsIsConstantValue(gutils, origops[1]) == 0 && API.EnzymeGradientUtilsIsConstantValue(gutils, orig) == 0
+
+    if API.EnzymeGradientUtilsIsConstantValue(gutils, origops[1]) == 0
```
Comment (vchuravy) [inline src/compiler.jl:3981]:
```suggestion
    if API.EnzymeGradientUtilsIsConstantValue(gutils, origops[1]) == 0 && API.EnzymeGradientUtilsIsConstantValue(gutils, orig) == 0
```

File: src/compiler.jl [hunk 5]
```diff
@@ -3840,26 +3954,27 @@ function arraycopy_common(fwd, B, orig, origArg, gutils, shadowdst)
     return nothing
 end
 
-function arraycopy_augfwd(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef, gutils::API.EnzymeGradientUtilsRef, normalR::Ptr{LLVM.API.LLVMValueRef}, shadowR::Ptr{LLVM.API.LLVMValueRef}, tapeR::Ptr{LLVM.API.LLVMValueRef})::Cvoid   
+function arraycopy_augfwd(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef, gutils::API.EnzymeGradientUtilsRef, normalR::Ptr{LLVM.API.LLVMValueRef}, shadowR::Ptr{LLVM.API.LLVMValueRef}, tapeR::Ptr{LLVM.API.LLVMValueRef})::Cvoid
     arraycopy_fwd(B, OrigCI, gutils, normalR, shadowR)
 
-    
+    shadowres = LLVM.Value(unsafe_load(shadowR))
```
Comment (wsmoses) [inline src/compiler.jl:3960]:
```suggestion
```

**Pull request merged**